src/faldbt/lib.py

```diff
@@ -101,6 +104,38 @@ def _get_adapter(
     return adapter
 
 
+global _lock
+# RLock supports recursive locking by the same thread
+_lock = threading.RLock()
+
+
+@contextmanager
+def _cache_lock(info: str = ""):
+    operationId = uuid4()
+    logger.debug("Locking  {} {}", operationId, info)
+
+    _exc = None
+    _lock.acquire()
+    logger.debug("Acquired {}", operationId)
+
+    try:
+        yield
+    except BaseException as exc:
+        _exc = exc
+
+    _lock.release()
+    logger.debug("Released {}", operationId)
+    if _exc:
+        raise RuntimeError(
+            f"Error during cache lock (operation: {operationId})"
+        ) from _exc
```

Comment:
Let's maybe just use try/finally here and propagate the exception directly (unless there is a reason to stick with `RuntimeError`, which is a behavioral change),

```py
    try:
        yield
    finally:
        _lock.release()
        logger.debug("Released {}", operationId)
```

Reply:
But it does work. Thanks!